transports/tcp/src/lib.rs

```diff
@@ -22,362 +22,639 @@
 //!
 //! # Usage
 //!
-//! This crate provides two structs, `TcpConfig` and `TokioTcpConfig`, depending on which
-//! features are enabled.
-//!
-//! Both the `TcpConfig` and `TokioTcpConfig` structs implement the `Transport` trait of the
-//! `core` library. See the documentation of `core` and of libp2p in general to learn how to
-//! use the `Transport` trait.
+//! This crate provides a `TcpConfig` and `TokioTcpConfig`, depending on
+//! the enabled features, which implement the `Transport` trait for use as a
+//! transport with `libp2p-core` or `libp2p-swarm`.
+
+mod provider;
+
+#[cfg(feature = "async-io")]
+pub use provider::async_io;
 
-use futures::{future::{self, Ready}, prelude::*};
+/// The type of a [`GenTcpConfig`] using the `async-io` implementation.
+#[cfg(feature = "async-io")]
+pub type TcpConfig = GenTcpConfig<async_io::Tcp>;
```

Comment:
```suggestion
pub type AsyncIoTcpConfig = GenTcpConfig<async_io::Tcp>;
```

As this pull request already introduces breaking changes, and thus an additional one should be less painful for users to handle, I would like to suggest renaming `TcpConfig` to `AsyncIoTcpConfig`. In my opinion `AsyncIoTcpConfig` is more expressive in regards to the fact that it is on par with `TokioTcpConfig`.

What do you think? As this is a bit unrelated to this pull request, I am happy to delay the discussion.

Reply:
I wanted to avoid further API changes, also considering that, as currently in `master`, the `async-std`/`async-io` variant is enabled by default, so using the unprefixed name for the configuration of the default feature seems justifiable.